default/python/universe_generation/empires.py

```diff
@@ -154,11 +169,15 @@ def find_home_systems_for_min_jump_distance(self, systems_pool, min_jumps):
                 # remove all neighbors from the local pool
                 local_pool -= set(get_systems_within_jumps(member, min_jumps))
 
-            # Count complete misses when number of candidates is not close to the target.
-            if len(candidate) < (self.num_home_systems - 4):
-                num_complete_misses_remaining -= 1
-
             if len(candidate) < self.num_home_systems:
+
+                # Break if the normal distribution of failed candidates is too much smaller than
+                # the target number of systems
+                lens_of_failed_candidates.append(len(candidate))
+                mean_len_of_failed_candidates = sum(lens_of_failed_candidates) / float(len(lens_of_failed_candidates))
+                if mean_len_of_failed_candidates < lowest_continuable_mean_len_of_candidates:
+                    break
```

Comment:
tl;dr;  I recommend going with the original solution of breaking immediately if `len(candidates)` is too small.


First, we should return to the original version of the PR and quit as soon as we fail, `len(candidate) < self.num_home_systems` and proceed to try an easier pool.

The algorithm tries to find a set of _m_target_ home systems out of _n_ total.  It first tries with a spacing _s_ between home systems of 10 and a series of more relaxed systems pools.  Then it repeats for a system spacing of 9 through 5.  For a given set of _n_ systems and spacing _s_ there is some _m_max_ home systems and if it is less than the _m_target_ the algorithm can't succeed. Trying more is a wasted effort.

When the algorithm fails it produces _m_failed_ candidates with some probability distribution.  It is only reasonable to try a second attempt if there is enough overlap of _m_target_ and this distribution to expect to succeed.  Otherwise, it is faster to try one of the more relaxed pools.

The variance of _m_failed_ is small relative to _n_.  This is easy to verify by printing out the `len(candidate)` and eyeballing it.  I provide some data below.  Consequently, if the first candidate is too small then all other candidates from the same pool with the same _s_ will with high probability be too small.

So actually using the variance to decide when to continue, only provides improved results if the first failed candidate is within some reasonable distance of _m_target_ home systems.  The total improvement is an increase in _s_ of 1.  This is a lot of angst for minimal gain, which is why I would return to the original PR code and quit immediately  and try an easier pool.

Let me try to explain the theory more clearly.

The `std = k * (_n_ systems)^(1/2)` for some coefficient _k_ is not an arbitrary interpolation.  Note, I made a mistake and carried a cubed root through all of the calculation.  It is expected standard deviation of _m_failed_home_systems_ from the math.  This can be shown in three steps that the distribution is expected to be normal, to have a mean proportional to _m_systems and have a standard deviation as above.

First, for a sufficiently large galaxy we could imagine it as _p_ galaxies stitched together.  We expect the big galaxy to be able to fit _p_ times as many home systems as each individual small galaxy.  Sufficiently large means that we ignore the perimeter effects.  So we expect the distribution of _m_failed_ to have a mean proportional to _n_.

Second, as our large galaxy is composed of _p_ smaller galaxies its _m_failed_ will be the sum of all the _p_ smaller galaxies _m_failed_ random variables.  We apply the central limit theorem, and know that such a sum of independent identically distributed random variables will have a normal distribution.  We now know that _m_failed_home_systems_ will be a normal random variable with a mean proportional to _n_ systems

Third and finally, also from the central limit theorem we know that the variance will be _p_ * _variance_small_galaxy_.  The standard deviation is the sqrt of the variance.  And again for some sufficiently large galaxy we can make _p_ proportional to _n_ systems.  Hence we get the final result `std = k * (_n_ systems)^(1/2)`.

So it only remains to get an estimate for _k_.  These were the experiments that I did.  As you pointed out, _k_ depends on the galaxy topology.  The 3 arm spiral std is about half of the cluster galaxy std.  However, we are estimating this variance in order to exclude cases that would run pointless trials, so I used the smaller variance.


Why not do this fitting in game?  It takes a long time for large galaxies, completely defeating the intent of using it to avoid wasteful computation.  Furthermore, we know the distribution a priori and it will not change in game.  Empirically determine the threshold and just use it in game. 


Here are some concrete numbers from a 3-arm spiral galaxy with 999 systems (1004 actually generated) targeting 39 home systems.  These are the length of the first 100 candidates for the first pool at a `min_jumps` of 10:

[13, 15, 16, 15, 16, 17, 16, 15, 17, 14, 15, 14, 14, 16, 17, 15, 16, 15, 14, 14, 15, 15, 13, 15, 15, 14, 17, 14, 16, 15, 15, 14, 14, 16, 16, 14, 17, 15, 13, 16, 15, 16, 16, 15, 16, 15, 17, 16, 16, 16, 17, 17, 14, 15, 15, 14, 15, 13, 16, 16, 16, 14, 18, 14, 16, 15, 15, 16, 16, 15, 17, 16, 15, 15, 14, 16, 17, 15, 15, 16, 15, 15, 13, 14, 15, 16, 15, 14, 17, 15, 14, 15, 15, 15, 15, 16, 16, 15, 13, 16]

You can verify that the mean is 15.23 and the standard deviation is 1.09. The goal is 23.85 standard deviations away from the first candidate, 13.  The next 99 trails are wasted computation.  There is no need to wait for an estimator to settle in order to decide to skip to the next trial.

In summary, I recommend removing the variance portion of the exit condition, and break if any candidate is shorter than the goal.  For galaxy up to the five times the current maximum size this is the correct decision.  Even for much larger galaxies the certainty of trying with an easier pool out-weighs the minimal benefit an increase of one in the minimum home system spacing.

This exit condition is useful, particularly if there is interest in large galaxies, because these extra trials are completely wasted computation and they scale worse than linear.